csrc/preseg_passes/remove_bcast_squeeze.cpp

```diff
@@ -10,44 +10,103 @@
 #include <logical_domain_map.h>
 #include <options.h>
 #include <preseg_passes/remove_bcast_squeeze.h>
-
 namespace nvfuser::preseg_passes {
 
 namespace {
 // Remove broadcast-squeeze and squeeze-broadcast patterns
-// TODO: still remove when have intermediate ops between broadcast and squeeze
+// Limitation: only support one producer and one consumer for tensors between
+// broadcast and squeeze
+// TODO: extend to allow multiple producers and consumers.
 void removeBcastSqueeze(Fusion* fusion) {
-  // Iterate backwards over fusion expressions.
-  // This will ensure we don't process expressions that are no longer valid
-  // after replacement.
-  auto exprs = fusion->exprs();
-  for (auto it = std::rbegin(exprs); it != std::rend(exprs); it++) {
-    Expr* expr = *it;
-    // step-1: find and remove broadcast + squeeze pattern
-    // before: Y = broadcast(X); Z = squeeze(Y);  M = someOp(Z)
-    // after : M = someOp(X)
-    // conditions: (1) broadcast and squeeze have the same dim flags
-    if (auto squeeze = dynamic_cast<SqueezeOp*>(expr)) {
-      if (auto bcast =
-              dynamic_cast<BroadcastOp*>(squeeze->in()->definition())) {
-        if (bcast->getBroadcastDimFlags() == squeeze->getSqueezeDimFlags()) {
-          ir_utils::replaceValInAllExprInputsAndFusionOutputs(
-              squeeze->out(), bcast->in());
+  bool is_removed = true;
+  // Loop over all exprs in fusion until no more patterns are removed
+  while (is_removed) {
+    // init the flag to false, if any pattern is removed, set it to true
+    is_removed = false;
+    // re-fetch the exprs after removing the patterns
+    auto exprs = fusion->exprs();
+    // Iterate backwards over fusion expressions.
+    // This will ensure we don't process expressions that are no longer valid
+    // after replacement.
+    for (auto it = std::rbegin(exprs); it != std::rend(exprs); it++) {
+      Expr* expr = *it;
+      // step-1: find and remove broadcast + squeeze pattern
+      // before: Y0 = broadcast(X); Yn = PointwiseOps(Y0), Z = squeeze(Yn);
+      //         M = someOp(Z)
+      // after : M = someOp(PointwiseOps(X))
+      // broadcast Ids are removed from all tensors between broadcast and
+      // squeeze
+      if (auto squeeze = dynamic_cast<SqueezeOp*>(expr)) {
+        auto bcast_tv = squeeze->in()->as<TensorView>();
+        // If this bcast tv has multiple consumers, don't remove the broadcast
+        // id
+        if (ir_utils::consumerTvsOf(bcast_tv).size() > 1) {
```

Comment:
Add a description of the approach you're taking. I'm concerned about this approach generally being a fix for the referenced issue. As it seems likely we could come up with an example that would reproduce a very similar pattern but this pass wouldn't be able to remove the broadcast.